### ISSUE TITLE
Node fails to start due to unreachable seed node

### ISSUE THREAD

**Issue by xujiaxj** [hekate-io/hekate#8]
I love this library, and have started to use it for our clustering need. Lately I noticed an issue that I couldn't find a workaround.

We are using JDBC as seed node provider. I understand each node adds itself to the database table `cluster_node` when it joins a cluster and removes itself from the table when it leaves the cluster. In addition, there is a `cleanup-interval` used by the coordinator node to remove inactive nodes.

Occasionally when all the nodes abruptly die together and fail to clean up the table, the table is left with dead nodes. If I start a new node, it tries to reach these dead "seed nodes" and falls into an infinite retry loop.

The exact sequence of events is non-trivial to reproduce, but the same situation can be easily simulated like the following.

Before start any node, manually add a fake node to the table to simulate a dead node. 
<img width="351" alt="image" src="https://user-images.githubusercontent.com/4541662/167516744-e181ea76-aeb5-45e4-844f-207e9a8bf1b3.png">

This fake node is unreachable, now start a real node `192.168.86.129`, and watch that it falls in an infinite loop trying to reach the unreachable node.

```
INFO  2022-05-09 23:38:56,956 main HekateNode 
      Hekate 4.0.0 (build: 20.03.2021, revision: 87b0f48e2d962d000b5c3108fec35e54e6c8b558).
INFO  2022-05-09 23:38:56,960 HekateSys-0-0 NettyNetworkService 
      Network address selected [address=/192.168.86.129, selector=any-ip4]
INFO  2022-05-09 23:38:56,961 HekateSys-0-0 NettyNetworkService 
      Binding network acceptor [port=10012]
INFO  2022-05-09 23:38:57,057 HekateNioAcceptor-2-0 NettyNetworkService 
      Couldn't bind on port 10012 ...will try next port [next-port=10013]
INFO  2022-05-09 23:38:57,061 HekateNioAcceptor-2-0 NettyNetworkService 
      Done binding [bind-address=/0:0:0:0:0:0:0:0:10013]
INFO  2022-05-09 23:38:57,083 HekateSys-0-0 NettyNetworkService 
      Started accepting network connections [address=/0:0:0:0:0:0:0:0:10013]
INFO  2022-05-09 23:38:57,085 HekateCluster-7-0 DefaultClusterService 
      Joining cluster [namespace=model-builder, local-node=/192.168.86.129:10013:f92ee540e26b405090441ebd39d81e75]
INFO  2022-05-09 23:38:57,086 HekateCluster-7-0 HikariDataSource 
      hekateDataSource - Starting...
INFO  2022-05-09 23:38:57,099 HekateCluster-7-0 HikariDataSource 
      hekateDataSource - Start completed.
INFO  2022-05-09 23:38:57,112 HekateClusterGossip-6-0 GossipSeedNodesSate 
      Discovered new seed node [address=SeedNodeState[address=/192.168.0.1:8080, status=NEW]]
INFO  2022-05-09 23:38:57,114 HekateClusterGossip-6-0 GossipSeedNodesSate 
      Discovered new seed node [address=SeedNodeState[address=/192.168.86.129:10012, status=NEW]]
INFO  2022-05-09 23:38:57,115 HekateClusterGossip-6-0 DefaultClusterService 
      Sending cluster join request [seed-node=/192.168.0.1:8080].
INFO  2022-05-09 23:38:58,118 HekateClusterGossip-6-0 DefaultClusterService 
      Sending cluster join request [seed-node=/192.168.86.129:10012].
INFO  2022-05-09 23:38:59,123 HekateClusterGossip-6-0 DefaultClusterService 
      Sending cluster join request [seed-node=/192.168.86.129:10012].
INFO  2022-05-09 23:39:00,130 HekateClusterGossip-6-0 DefaultClusterService 
      Sending cluster join request [seed-node=/192.168.86.129:10012].
WARN  2022-05-09 23:39:00,150 HekateClusterGossip-6-0 GossipSeedNodesSate 
      Seed node timeout ...will retry [address=/192.168.0.1:8080, cause=io.hekate.network.NetworkConnectTimeoutException: connection timed out: /192.168.0.1:8080]
WARN  2022-05-09 23:39:01,124 HekateClusterGossip-6-0 GossipSeedNodesSate 
      Seed node timeout ...will retry [address=/192.168.0.1:8080, cause=io.hekate.network.NetworkConnectTimeoutException: connection timed out: /192.168.0.1:8080]
INFO  2022-05-09 23:39:01,134 HekateClusterGossip-6-0 DefaultClusterService 
      Sending cluster join request [seed-node=/192.168.86.129:10012].
WARN  2022-05-09 23:39:02,128 HekateClusterGossip-6-0 GossipSeedNodesSate 
      Seed node timeout ...will retry [address=/192.168.0.1:8080, cause=io.hekate.network.NetworkConnectTimeoutException: connection timed out: /192.168.0.1:8080]
INFO  2022-05-09 23:39:02,169 HekateClusterGossip-6-0 DefaultClusterService 
      Sending cluster join request [seed-node=/192.168.0.1:8080].
WARN  2022-05-09 23:39:03,132 HekateClusterGossip-6-0 GossipSeedNodesSate 
      Seed node timeout ...will retry [address=/192.168.0.1:8080, cause=io.hekate.network.NetworkConnectTimeoutException: connection timed out: /192.168.0.1:8080]
WARN  2022-05-09 23:39:03,154 HekateClusterGossip-6-0 GossipSeedNodesSate 
      Seed node timeout ...will retry [address=/192.168.0.1:8080, cause=io.hekate.network.NetworkConnectTimeoutException: connection timed out: /192.168.0.1:8080]
```

**Comment by syermakov** [comment]
Hi @xujiaxj

Thank you for the detailed problem explanation and for providing logs.

I can see from the logs that there are network timeouts when `192.168.86.129` tries to contact `192.168.0.1`:
```
Seed node timeout ...will retry [address=/192.168.0.1:8080, cause=io.hekate.network.NetworkConnectTimeoutException: connection timed out: /192.168.0.1:8080]
```

In case of network timeouts we can't really judge on whether the seed node is really dead (could be simply paused by GC) or is it still alive. If we would ignore such errors and continue joining the cluster then it could potentially lead to cluster split-brain issue.

Do you see the same type of errors in the real scenario when multiple nodes abruptly die together?  
Also, could it be some kind of network/firewall configuration issue that caused network timeouts instead of connection being immediately refused?

**Comment by xujiaxj** [comment]
Yes. We've seen the same type of error messages in our real cluster.
<img width="1210" alt="image" src="https://user-images.githubusercontent.com/4541662/167730390-b6923105-3652-4789-94b8-534f273de15f.png">

We've confirmed these nodes are old nodes that are either scaled down or went dead, but somehow didn't clean themselves from the table. It's not an issue of connectivity.

**Comment by syermakov** [comment]
I'm not sure why it results in network timeout errors instead of immediate connection failures. Sometimes it can happen because of security groups configuration or could be something environment-specific, but anyway it out of our control at the library level.

I suggest to try version 4.1.1 (just released). 

It has a configuration flag to control handling of seed node timeouts:
 - In Spring Boot: `hekate.cluster.seed-node-fail-fast`
 - Programmatically: `ClusterServiceFactory#setSeedNodeFailFast(boolean)`

If this flag is set to `true`, then network timeout errors will be treated as permanent failures and the target node will be removed from the list of active seed nodes.

I would also suggest to set the network connect timeout to a higher value (by default it is 3 seconds) when `seed-node-fail-fast` is set to `true`.  It will help to mitigate potential cluster split issues when a new node joins the cluster while the seed nodes are hanged for whatever reasons. Connect timeout can be set via `hekate.network.connect-timeout` Spring Boot property or programmatically via `NetworkServiceFactory#setConnectTimeout(int)` method.

**Comment by xujiaxj** [comment]
Thanks for implementing the flags. Is 4.1.1 going to be available to pull from maven central any time soon?

Meanwhile, I can definitely see something inconsistent on how the library behave. If I change my fake node from `192.168.0.1` to `193.168.0.1` to reverse the order of the seed nodes, my local node `192.168.86.129` can start successfully. It hit the timeout exception twice and moved on.

```
INFO  2022-05-11 16:03:25,122 main HekateNode 
      Hekate 4.1.0 (build: 03.04.2022, revision: 90aef2e3a208bdef5c2afec8ad08634a5fef13b0).
INFO  2022-05-11 16:03:25,127 HekateSys-0-0 NettyNetworkService 
      Network address selected [address=/192.168.86.129, selector=any-ip4]
INFO  2022-05-11 16:03:25,127 HekateSys-0-0 NettyNetworkService 
      Binding network acceptor [port=10012]
INFO  2022-05-11 16:03:25,249 HekateNioAcceptor-2-0 NettyNetworkService 
      Done binding [bind-address=/0:0:0:0:0:0:0:0:10012]
INFO  2022-05-11 16:03:25,279 HekateSys-0-0 NettyNetworkService 
      Started accepting network connections [address=/0:0:0:0:0:0:0:0:10012]
INFO  2022-05-11 16:03:25,282 HekateCluster-7-0 DefaultClusterService 
      Joining cluster [namespace=model-builder, local-node=/192.168.86.129:10012:d7eeb03e5a35492f97df9d92d99f8c55]
INFO  2022-05-11 16:03:25,283 HekateCluster-7-0 HikariDataSource 
      hekateDataSource - Starting...
INFO  2022-05-11 16:03:25,297 HekateCluster-7-0 HikariDataSource 
      hekateDataSource - Start completed.
INFO  2022-05-11 16:03:25,309 HekateClusterGossip-6-0 GossipSeedNodesSate 
      Discovered new seed node [address=SeedNodeState[address=/193.168.0.1:8080, status=NEW]]
INFO  2022-05-11 16:03:25,311 HekateClusterGossip-6-0 DefaultClusterService 
      Sending cluster join request [seed-node=/193.168.0.1:8080].
INFO  2022-05-11 16:03:26,315 HekateClusterGossip-6-0 DefaultClusterService 
      Sending cluster join request [seed-node=/193.168.0.1:8080].
INFO  2022-05-11 16:03:27,323 HekateClusterGossip-6-0 DefaultClusterService 
      Sending cluster join request [seed-node=/193.168.0.1:8080].
INFO  2022-05-11 16:03:28,330 HekateClusterGossip-6-0 DefaultClusterService 
      Sending cluster join request [seed-node=/193.168.0.1:8080].
WARN  2022-05-11 16:03:28,343 HekateClusterGossip-6-0 GossipSeedNodesSate 
      Seed node timeout ...will retry [address=/193.168.0.1:8080, cause=io.hekate.network.NetworkConnectTimeoutException: connection timed out: /193.168.0.1:8080]
WARN  2022-05-11 16:03:29,318 HekateClusterGossip-6-0 GossipSeedNodesSate 
      Seed node timeout ...will retry [address=/193.168.0.1:8080, cause=io.hekate.network.NetworkConnectTimeoutException: connection timed out: /193.168.0.1:8080]
INFO  2022-05-11 16:03:29,339 HekateSys-0-0 HekateNode 
      Joined the cluster ...will synchronize [join-order=1, topology=ClusterTopology[size=1, version=1, nodes=[/192.168.86.129:10012:d7eeb03e5a35492f97df9d92d99f8c55]]]
INFO  2022-05-11 16:03:29,344 HekateClusterEvent-1-0 ClusterEventLogger 
      
*** Join topology [size=1, version=1] {
         this -> [address=/192.168.86.129:10012:d7eeb03e5a35492f97df9d92d99f8c55, name=, roles=[], properties={}, runtime=ClusterNodeRuntime[cpus=16, mem=7.11Gb, os=Mac OS X, jvm=1.8.0_242, pid=34450], join-order=1]
    }
INFO  2022-05-11 16:03:29,349 HekateSys-0-0 HekateNode 
      Hekate is UP and running [node=/192.168.86.129:10012:d7eeb03e5a35492f97df9d92d99f8c55]
```

So in summary, my local node is `192.168.86.129`, and if the seed node list is ordered as [`192.168.0.1`, `192.168.86.129`], my local node falls into an infinite loop retrying when starting up. If the seed node list is ordered as [`192.168.86.129`, `193.168.0.1`], my local node can start fine after two timeouts.

**Comment by syermakov** [comment]
> Is 4.1.1 going to be available to pull from maven central any time soon?

It is already available: https://repo1.maven.org/maven2/io/hekate/hekate-core/4.1.1/

> If the seed node list is ordered as [192.168.86.129, 193.168.0.1], my local node can start fine after two timeouts.

I'll look into it.

**Comment by syermakov** [comment]
@xujiaxj 
I was able to reproduce the seed node order issue. 
I'll fix it and prepare a new release (4.1.2) shortly.

**Comment by xujiaxj** [comment]
Great! Thx for the fast turnaround!

**Comment by syermakov** [comment]
v.4.1.2 is available from Maven Central. Please give it try.

**Comment by xujiaxj** [comment]
I tested 4.1.2, it doesn't fix the infinite retry, but it makes the infinite retry behavior consistent, whether the fake node is `192.168.0.1` or `193.168.0.1`. The new flag `hekate.cluster.seed-node-fail-fast` still works. Is this what you would expect the fix to do?

**Comment by syermakov** [comment]
Yes, expected behavior is:

 - if `fail-fast = false` then it should always retry timeout errors
 - if `fail-fast = true` then it should stop retrying timeout errors and should blacklist such address

**Comment by xujiaxj** [comment]
kk. We can close it for now. Again thanks for the help!
